File: merge_sort/merge.js
// Algorithms, 4th Edition by Robert Sedgewick
// implementation of book algorithms in JavaScript
// author of javascript implementation Steven Aguilar
// Githut: @theasteve  email: stevenaguilarcs@gmail.com


function merge(a, lo, mid, hi) {
  let aux = []
  let i = lo
  let j = mid + 1

  for(let i = lo; i <= hi; i++) {
    aux[i] = a[i]
  }
  for(let k = lo; k <= hi; k++) {
    if(i > mid) {
      a[k] = aux[j++]
    }
    else if(j > hi) {
      a[k] = aux[i++]
    }
    else if(aux[j] < aux[i]){
      a[k] = aux[j++]
    }
    else{
      a[k] = aux[i++]
    }
  }
  console.log(a)
  return a;
}

a = ['E','E','G','M','R','A','C','E','R','T']
lo = 0
mid = parseInt((a.length - 1) / 2)
hi = a.length - 1

merge(a,lo,mid,hi)

File: merge_sort/merge.rb
# Algorithms, 4th Edition by Robert Sedgewick
# implementation of book algorithms in JavaScript
# author of Ruby implementation Steven Aguilar
# Githut: @theasteve  email: stevenaguilarcs@gmail.com

require 'pry'

def merge(a, lo, mid, hi)
  i = lo
  j = mid + 1
  aux = []

  k = lo
  while k <= hi do
    aux[k] = a[k]
    k += 1
  end

  k = lo
  while k <= hi do
    if i > mid
      a[k] = aux[j]
      j += 1
    elsif j > hi
      a[k] = aux[i]
      i += 1
    elsif aux[j] < aux[i]
      a[k] = aux[j]
      j += 1
    else
      a[k] = aux[i]
      i += 1
    end

    k += 1
  end

  a
end

a = ['E','E','G','M','R','A','C','E','R','T']
lo = 0
mid = (a.length - 1) / 2
hi = a.length - 1

p merge(a, lo, mid, hi)
